검색결과 화면 UI 및 라우팅

KimJye opened the pull request:
## 작업 내용
검색 결과 화면을 구현했습니다.
검색 결과 없음을 구현했습니다.

## 실행 방법
검색입력창에 검색어를 입력한다.
여러개의 탭을 눌러본다~

## 결과 첨부

![image](https://user-images.githubusercontent.com/33201628/70290348-eba69280-181a-11ea-8387-67a30b13b3a4.png)

![image](https://user-images.githubusercontent.com/33201628/70290360-f8c38180-181a-11ea-8de9-63f81673dfa6.png)

![결과화면없음](https://user-images.githubusercontent.com/33201628/70290310-d03b8780-181a-11ea-9d83-5aa98309becb.png)

KimJye committed:
[client] | feat(search): 예전에 만들어 놓은 검색 결과로 작성해봄

더미데이터 넣었습니다.

KimJye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/searched_ui_20191203

KimJye committed:
[client] | feat(search): 검색 결과화면 모바일 마크업

모바일 기준으로 검색 결과화면을 마크업했습니다.

KimJye committed:
[client] | feat(search): 데스크톱버전을 위해서 가운데로 정렬함

KimJye committed:
[clinet] | feat(searched): 검색결과 데스크톱 화면 마크업

미디어쿼리를 이용해서 구현했습니다

KimJye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/searched_ui_20191203

KimJye committed:
[client] | feat(search): custom hook 생성

fetching을하여 가져온 데이터로 결과를 보여줍니다.
query대신에 keyword로 변경하였습니다.
검색결과 처음에 보여지는 페이지를 분리할 예정입니다. 따라서 현재 구현한
페이지는 검색결과에서 모두에 해당하는 페이지입니다.

KimJye committed:
[client] | feat(search): 검색 탭별로 페이지 생성

videos, users, tags 각각의 탭별로 페이지를 생성했습니다.

KimJye committed:
[client] | refactor(search): 코드 중복을 피하기위해 함수로 분리

KimJye committed:
[client] | feat(search): 각 탭별로 제목 추가

KimJye committed:
[client] | fix(search): 인피니티 스크롤 관련 에러 수정

데이터가 있는지 확인을 하는 상태가 없어서 계속 스크롤에서 요청을 하는
에러가 있었습니다. 그래서 페이지가 제대로 맞지 않았었습니다.
따라서 hooks에 hasData라는 상태를 관리하였습니다.

KimJye committed:
[client] | refactor(search): 코드 중복최소화를 위한 컴포넌트화

video, user, tag 각각의 결과 리스트를 컴포넌트화하였습니다.

KimJye committed:
[client] | refactor(search): 검색 결과 제목 컴포넌트화

여러곳에서 쓰이는 검색어 정보를 컴포넌트화하였습니다.

KimJye committed:
[client] | refactor(search): 더보기 버튼을 컴포넌트로 만듦

KimJye committed:
[client] | refactor(search): 파일 구조 변경

KimJye committed:
[client] | feat(search): 검색 결과에 따른 동적 탭 구현

각 주제에따라 탭 유무를 나타내도록 구현했습니다.

KimJye committed:
[client] | feat(search): 검색 결과 없는 화면 구현

검색 결과가 없을 때 화면을 구현했습니다.

KimJye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/searched_ui_20191203

KimJye committed:
[client] | refactor(search): 전체 보기 버튼에 대한 조건 수정

검색결과에대해서 모두 탭이 아닌 다른탭일때는 전체 보기 버튼이 없어야하기
때문에 코드를 수정했습니다.

KimJye committed:
[client] | feat(search): 검색결과 탭바 상태 관리

서버사이드렌더링을 위해서, 검색결과에 따른 탭바 상태를 query string으로
넘기는 작업을 하였습니다.
또한 '모두'라는 탭은 각 주제가 2개이상있을때 보여지도록 설정했습니다.

KimJye committed:
[client] | refactor(search): 검색결과없음에대한 코드 수정

검색결과는 여러개의 패칭을 하는데, 각각의 패칭의 결과에대해서 데이터가
없으면 검색결과가 없다고 했었습니다.
이에 문제는, 검색결과가 있는데 중간에 검색결과가 없다고
나오는것이었습니다.
이것을 해결하기 위해서 데이터 패칭이 다 끝났을때를 체킹합니다.

KimJye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/searched_ui_20191203

Atercatus commented on packages/client/components/SearchBar/index.tsx:
```diff
@@ -20,7 +20,7 @@ export const SearchBar: React.FunctionComponent<SearchBarProps> = ({
       router.push({
         pathname: endpoint.search,
-        query: { query: e.target.value },
+        query: { keyword: e.target.value },
```
state 를 쓰는게 더 좋지 않을까요?

riverandeye commented on packages/client/views/Search/index.tsx:
```diff
@@ -4,22 +4,142 @@ import { useRouter } from 'next/router';
+            {allCount === 0 || !hasData ? (
+              <S.NoResults>
+                <span>"{searchKeyword}"에 대한 검색결과가 없습니다...</span>
+              </S.NoResults>
```
아직 검색이 완료되지 않은 경우에도 검색결과가 없습니다... 라는 글귀가 나타나는 것으로 보입니다.
검색 기능 자체에 지장은 없으나, 사용성 측면에서 부적절해보여서 리뷰를 남깁니다.

hyokualexkwon replied:
로딩중에는 본 문구가 나타나지 않도록 변경해야 할 것 같습니다.

KimJye replied:
시정했습니다 감사합니다

riverandeye commented on packages/client/views/Search/hooks.ts:
```diff
@@ -0,0 +1,77 @@
+
+  const action = createSearchAction(search.users, page, keyword);
+  const { payload, error } = useQuery(action);
+
```
검색에 아무 값도 넣지 않는 경우에도 검색이 가능한 것으로 보입니다.
또, keyword 문자열을 trim 해서 공백을 제거해야 할 것 같이 보입니다. 
이 부분은 정말 복잡하고 어렵네요..

![image](https://user-images.githubusercontent.com/41494099/70393577-c1c5b980-1a2e-11ea-9d6a-a683791366f5.png)

KimJye replied:
호옷~~~ 감사합니다~!

riverandeye commented on packages/client/views/Search/index.tsx:
```diff
@@ -4,22 +4,142 @@ import { useRouter } from 'next/router';
+import { endpoint, searchOptions, search } from '../../constants';
+import { useSearchVideos, useSearchUsers, useSearchTags } from './hooks';
 
 const Searched: React.FunctionComponent = () => {
```
Searched 라는 이름이 조금 모호한 것 같아요. 
SearchResult 는 어떠신가요?

hyokualexkwon replied:
디렉토리명은 Search 인데 컴포넌트 이름은 Searched네요. 통일해야 할 것 같습니다.

riverandeye commented on packages/client/views/Search/index.tsx:
```diff
@@ -4,22 +4,142 @@ import { useRouter } from 'next/router';
   const router = useRouter();
+  const searchKeyword = router.query.keyword;
+
+  const page = undefined;
```
const page = null 이 더 적합한 것 같습니다. undefined는 조금 명확하지 않다고 생각이 듭니다.

riverandeye reviewed (changes requested):
어려운 기능 맡으신것 같아요 고생이 많으십니다.

hyokualexkwon commented on packages/client/components/Searched/ViewMore/index.tsx:
```diff
@@ -0,0 +1,29 @@
+        <a>
+          <button>
+            <span>전체 {searchOptions[num].label}</span>
+            <ArrowRightSVG />
```
화면에서는 svg의 opacity가 적용되어 있지 않네요!

hyokualexkwon commented on packages/client/components/Searched/Tags/index.tsx:
```diff
@@ -0,0 +1,21 @@
+        {tags.map(tag => {
+          return (
+            <div key={tag.id}>
+              <S.Tag>{tag.name}</S.Tag>;
```
`;`은 지우셔야겠네요

hyokualexkwon commented on packages/client/views/Search/Tags/index.tsx:
```diff
@@ -0,0 +1,105 @@
+  optionMap.set('all', '모두');
+  optionMap.set('videos', '영상');
+  optionMap.set('users', '사용자');
+  optionMap.set('tags', '태그');
```
컴포넌트가 렌더링 될 때마다 Map객체를 생성하고, 다수의 set을 하게 될것 같아요. 차라리 contstants에 별도의 객체 상수로 가지고 있는게 낫지 않을까요?

hyokualexkwon commented on packages/client/views/Search/Tags/index.tsx:
```diff
@@ -0,0 +1,105 @@
+    setPage(page + 1);
+  };
+
+  const routerObject = (queryKeyword, num) => {
```
라우터에 사용할 객체를 만드는 함수니까, 함수명 맨 앞에는 동사를 작성해야 할 것 같습니다. 동시에 자료형인 object도 생략해도 되겠네요.

hyokualexkwon commented on packages/client/views/Search/Videos/index.tsx:
```diff
@@ -0,0 +1,107 @@
+
+  const handleFilterClick = value => {
+    setPage(1);
+    if (value === searchOptions[0].value) {
```
searchOption 객체가 배열로 되어있는데, 특정 항목을 index로 접근하는 코드가 가독성이 저하되는것 같아요. 배열 대신에 객체를 활용해서 특정 key를 통해 값을 접근할 수 있게 한다면 이 부분의 코드 가독성이 향상되지 않을까요?

hyokualexkwon commented on packages/client/views/Search/index.tsx:
```diff
@@ -4,22 +4,142 @@ import { useRouter } from 'next/router';
+                  activeValue={activeSearch}
+                  onClick={handleFilterClick}
+                />
+                <S.Line />
```
구분자의 높이가 1px이 되도록 수정해 주세요. Video페이지에서 예시를 찾아볼 수 있습니다.

hyokualexkwon commented on packages/client/views/Search/styles.tsx:
```diff
@@ -39,43 +21,37 @@ export const ContainerGrid = styled(MaterialGrid)`
-  }
+export const Line = styled.hr`
+  border: 1px solid #484c50;
+`;
```
모바일 화면에서는 양 옆의 간격이 없어야 합니다. 비디오 페이지를 참고 해 주세요.

![image](https://user-images.githubusercontent.com/17868599/70682815-640dc780-1ce3-11ea-8ec1-f6edfc360339.png)

hyokualexkwon reviewed (commented):
전반적으로 좋습니다.

그런데 디렉토리 구조를 저희 프로젝트에 맞게 수정해야 할 것 같습니다. 컴포넌트 디렉토리는 또다른 컴포넌트 디렉토리의 sub 디렉토리가 되지 않도록 수정해야 합니다. 전부 components로 분리해 주세요. 물론 이름도 조금 바뀌어야겠죠?

KimJye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/searched_ui_20191203

KimJye committed:
[client] | refactor(search view): views의 디렉토리 변경

view 디렉토리는 또다른 디렉토리의 sub 디렉토리가 되지 않도록
수정했습니다.

KimJye committed:
[client] | refactor(search): 파일명 변경 및 컴포넌트화

중복 코드를 제거하고자 props를 이용하여 컴포넌트화 했습니다.
또한 기존 컴포넌트 규칙과 맞게 적용했습니다.

KimJye committed:
[client] | refactor(search): 검색 결과 부분 리팩토링

KimJye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/searched_ui_20191203

KimJye committed:
[client] | fix(search): 데이터로딩중 검색결과없음 나오는 에러 수정

null + null + null 했을때 0이 되는 문제가 있었습니다.
그래서 hasData가 true일때 기준으로 덧셈을 했고 false이면 null을
하였습니다.

KimJye committed:
[client] | fix(search): 검색 빈칸 없애기

KimJye committed:
[client] | refactor(searchBar): 검색 공백 에러 수정

공백이 있을때 input을 제어하도록 변경했습니다.
state로 관리했습니다.

KimJye committed:
[client] | fix(searched): 검색결과 없음 에러 처리

검색 결과를 로딩하는 동안과 결과가 없음에대한 에러가 있어서 수정했습니다.

KimJye committed:
[client] | refactor(searched): 검색결과화면에서 구분선 수정

구분선의 두께를 1로 변경했습니다.
모바일일때는 구분선이 화면에 꽉차도록 수정했습니다.

Atercatus commented on packages/client/components/SearchBar/index.tsx:
```diff
@@ -16,11 +16,20 @@ export const SearchBar: React.FunctionComponent<SearchBarProps> = ({
 }) => {
   const router = useRouter();
-  const sendQuery = e => {
+  const [value, setValue] = useState('');
```
value가 뭔지 알 기가 힘들어요 명확한 이름이면 좋겠어요

Atercatus commented on packages/client/components/SearchedArea/index.tsx:
```diff
@@ -0,0 +1,53 @@
+const SearchedArea = ({ subject, data, type }) => {
+  let children;
+
+  if (type === 'tags') {
```
문자열은 constants로 빼면 좋을 거 같습니다.

Atercatus commented on packages/client/components/SearchedArea/index.tsx:
```diff
@@ -0,0 +1,53 @@
+    });
+  }
+
+  if (type === 'users') {
```
문자열은 constants로 빼면 좋을 거 같습니다.

Atercatus commented on packages/client/components/SearchedArea/index.tsx:
```diff
@@ -0,0 +1,53 @@
+    });
+  }
+
+  if (type === 'videos') {
```
문자열은 constants로 빼면 좋을 거 같습니다.

Atercatus commented on packages/client/components/ViewMore/index.tsx:
```diff
@@ -0,0 +1,29 @@
+import { ArrowRightSVG } from '../../svgs';
+import { endpoint, searchOptions } from '../../constants';
+
+const ViewMore = ({ searchKeyword, num, options }) => {
```
num 이라는 변수는 다른 명확한 이름을 가진 변수로 바꾸면 좋을 거 같습니다.

Atercatus commented on packages/client/pages/search/index.tsx:
```diff
@@ -1 +1 @@
-export { default } from '../../views/Search';
+export { default } from '../../views/Searched';
```
Searched 는 명사가 아니라서 다른 이름이면 더 좋을 거 같아요.

Atercatus commented on packages/client/views/Searched/index.tsx:
```diff
@@ -0,0 +1,146 @@
+
+  const hasData = videoHasData && userHasData && tagHasData;
+
+  const allCount = hasData ? videoCount + userCount + tagCount : null;
```
비교를 숫자랑 하시던데 `null` 보단 `0` 이 더 명확할 거 같네요.

Atercatus commented on packages/client/views/Searched/index.tsx:
```diff
@@ -0,0 +1,146 @@
+  });
+
+  const handleFilterClick = value => {
+    if (value === searchOptions[1].value) {
```
무슨 조건인지 알기가 어려워요

Atercatus replied:
searchOptions를 배열로 하신이유가 있나요?

Atercatus commented on packages/client/views/Searched/index.tsx:
```diff
@@ -0,0 +1,146 @@
+                />
+                <S.Line />
+
+                {videoCount > 0 && (
```
`vidoCount &&` 로만 하면 안되나요?

Atercatus commented on packages/client/views/Searched/index.tsx:
```diff
@@ -0,0 +1,146 @@
+          <Grid item xs={12} md={8}>
+            <SearchedTitle searchKeyword={searchKeyword} />
+
+            {allCount === 0 && hasData ? (
```
무슨 조건인지 함수로 빼면 좋을 거 같습니다.

KimJye committed:
[client] | feat(searched): 검색 결과들 상세정보 라우팅

Link를 이용해서 라우팅을 하였습니다.

KimJye committed:
[clinet] | refactor(): 파일명 변경

searched -> searchedResults로 변경했습니다.

KimJye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/searched_ui_20191203

KimJye committed:
[client] | refactor(): api server host -> api url host

Atercatus commented on packages/client/views/SearchedResults/index.tsx:
```diff
@@ -0,0 +1,146 @@
+    .join(',');
+
+  const activeSearch =
+    customSearchOptions[0] === undefined ? 'all' : customSearchOptions[0].value;
```
`customSearchOptions[0] ? 'all' : customSearchOptions[0].value;` 로 하지 않으신 이유가 있나요

Atercatus commented on packages/client/components/SearchedResultsArea/index.tsx:
```diff
@@ -0,0 +1,62 @@
+const SearchedResultsArea = ({ subject, data, type }) => {
+  let children;
+
+  if (type === 'tags') {
```
문자열은 상수로 빼주세요

Atercatus commented on packages/client/views/SearchedResults/hooks.ts:
```diff
@@ -0,0 +1,77 @@
+  const { payload, error } = useQuery(action);
+
+  useEffect(() => {
+    if (payload && !error) {
```
early return 을 사용해주세요

Atercatus commented on packages/client/views/SearchedResults/hooks.ts:
```diff
@@ -0,0 +1,77 @@
+  const { payload, error } = useQuery(action);
+
+  useEffect(() => {
+    if (payload && !error) {
```
early return 을 사용해 주세요

Atercatus commented on packages/client/views/SearchedResults/index.tsx:
```diff
@@ -0,0 +1,146 @@
+  });
+
+  const handleFilterClick = value => {
+    if (value === searchOptions[1].value) {
```
무슨 조건인 지 알수 있게 명확한 함수명을 가진 함수로 분리하는게 좋을 거 같습니다.

Atercatus commented on packages/client/views/SearchedResults/index.tsx:
```diff
@@ -0,0 +1,146 @@
+    query: { keyword: queryKeyword, options },
+  });
+
+  const handleFilterClick = value => {
```
value가 어떤 value 인지 알기 힘듭니다.

Atercatus commented on packages/client/views/SearchedResults/index.tsx:
```diff
@@ -0,0 +1,146 @@
+    if (value === searchOptions[1].value) {
+      router.push(routerObject(searchKeyword, 1));
+    }
+    if (value === searchOptions[2].value) {
```
serachOptions 를 배열로 만드신 특별한 이유가 있나요?

Atercatus commented on packages/client/views/SearchedResults/index.tsx:
```diff
@@ -0,0 +1,146 @@
+    { label: '태그', value: 'tags', count: tagCount },
+  ];
+
+  const customSearchOptions = countArray.reduce((acc, cur) => {
```
map으로 대체하면 더 좋을 거 같습니다.

Atercatus commented on packages/client/views/SearchedResultsTags/index.tsx:
```diff
@@ -0,0 +1,106 @@
+
+  const [page, setPage] = useState(1);
+
+  const activeSearch = searchOptions[3].value;
```
어떤 값인지 알 기 힘듭니다.

Atercatus commented on packages/client/views/SearchedResultsTags/index.tsx:
```diff
@@ -0,0 +1,106 @@
+      };
+    }
+    return {
+      pathname: `${endpoint.search}/${searchOptions[num].value}`,
```
`searchOptions[num].value` 와 같은 값은 어떤 num의 어떤 value 인지 알 기 힘듭니다.

Atercatus commented on packages/client/views/SearchedResultsTags/index.tsx:
```diff
@@ -0,0 +1,106 @@
+    };
+  };
+
+  const handleFilterClick = value => {
```
코드의 중복이 많고 조건이 너무 많습니다.

Atercatus commented on packages/client/views/SearchedResultsVideos/index.tsx:
```diff
@@ -0,0 +1,112 @@
+  optionMap.set('users', '사용자');
+  optionMap.set('tags', '태그');
+
+  const customSearchOptions = optionArray.reduce((acc, cur) => {
```
map이 더 적합할 거 같습니다.

Atercatus commented on packages/client/views/SearchedResultsVideos/index.tsx:
```diff
@@ -0,0 +1,112 @@
+  };
+
+  const routerObject = (queryKeyword, num) => {
+    if (num === 0) {
```
`if(num)` 으로 해서 ㅇ반대로 하면 좋을 거 같습니다.

Atercatus commented on packages/client/views/SearchedResultsVideos/index.tsx:
```diff
@@ -0,0 +1,112 @@
+
+  const handleFilterClick = value => {
+    setPage(1);
+    if (value === searchOptions[0].value) {
```
중복된 코드가 많고 조건이 많습니다. 그리고 0과 1  2 3 이 각각이 무엇을 뜻하는 지 알 수 가 없습니다.

Atercatus reviewed (changes requested):
많은 코드를 작성하신다고 고생 많으셨습니다.
중복된 코드를 줄이고 조건 변경에 예민한 조건문들을 삭제하는 것 이 좋을 거 같습니다.
또한 명확하지 않은 변수명은 가독성을 떨어트려 리뷰어에게 많은 부담이 되므로 피하시는게 좋을 거 같습니다.

KimJye committed:
[client] | refactor(): 상수로 변경

KimJye committed:
[client] | refactor(): state의 이름 변경

더 명확하게 표현하기 위해서 value에서 inputValue로 변경하였습니다.

KimJye committed:
[client] | refactor(veiwMore): 변수이름 변경

hyokualexkwon commented on packages/client/components/SearchedResultsArea/styles.tsx:
```diff
@@ -0,0 +1,60 @@
+import styled from 'styled-components';
```
![image](https://user-images.githubusercontent.com/17868599/70775880-f20ab080-1dbe-11ea-8631-a69a0cf16b0a.png)

검색 결과의 각 항목의 선택 가능영역을 컨텐츠 영역으로 제한해야 합니다. 빈공간을 클릭해도  a태그를 누른 효과가 나타나면 안되겠네요.

hyokualexkwon commented on packages/client/components/SearchedResultsArea/styles.tsx:
```diff
@@ -0,0 +1,60 @@
+import styled from 'styled-components';
+import { BREAKPOINT } from '../../constants';
```
![image](https://user-images.githubusercontent.com/17868599/70776003-3f871d80-1dbf-11ea-9891-9fa8cb26cd37.png)

안삐져나가게 해주세요.
코드리뷰 제출전에 기본적인 ui 체크는 해야겠죠?

hyokualexkwon commented on packages/client/views/SearchedResultsTags/index.tsx:
```diff
@@ -0,0 +1,106 @@
+  const router = useRouter();
+  const searchKeyword = router.query.keyword;
+  const options = router.query.options as string;
+  const optionArray = options.split(',');
```
UI의 상태를 url 쿼리스트링으로 관리하는건 말이 안됩니다. 상태는 자바스크립트로 관리해 주세요.

`http://localhost:3000/search/videos?keyword=java&options=all%2Cvideos%2Ctags`

hyokualexkwon commented on packages/client/views/SearchedResultsUsers/index.tsx:
```diff
@@ -0,0 +1,113 @@
+    };
+  };
+
+  const handleFilterClick = value => {
```
본 메소드에서 반복적으로 작성되는 코드를 줄일 수 있을 것 같습니다.

hyokualexkwon commented on packages/client/components/ViewMore/index.tsx:
```diff
@@ -0,0 +1,29 @@
+      <Link
+        prefetch={false}
+        href={{
+          pathname: `${endpoint.search}/${searchOptions[num].value}`,
```
음 아무래도 배열에 index로 접근하는 것은 가독성이 많이 떨어지네요. 새로운 객체 상수를 만들어야 할 것 같습니다.

hyokualexkwon commented on packages/client/constants.ts:
```diff
@@ -92,6 +92,31 @@ export const signUpFormDataMaxLength = {
+  tags: 'tags',
+  all: 'all',
+};
+export const searchOptions = [
```
```
const searchOptionLabels = {
  all: '모두',
  videos: '영상',
  users: '사용자',
  tags: '태그'
}
```

```
export const searchOptionValues = {
  all: 'all',
  videos: 'videos',
  users: 'users',
  tags: 'tags'
};
```

```
export const searchOptions = [
  {
    label: 'searchOptionLabels.all',
    value: searchOptionValues.all,
  },
  {
    label: 'searchOptionLabels.videos',
    value: searchOptionValues.videos,
  },
  {
    label: 'searchOptionLabels.users',
    value: searchOptionValues.users,
  },
  {
    label: 'searchOptionLabels.tags',
    value: searchOptionValues.tags,
  },
]
```

이런 상수 관리는 어떨까요?

hyokualexkwon commented on packages/client/views/SearchedResults/index.tsx:
```diff
@@ -0,0 +1,146 @@
+import { endpoint, searchOptions, search } from '../../constants';
+import { useSearchVideos, useSearchUsers, useSearchTags } from './hooks';
+
+const SearchedResults: React.FunctionComponent = () => {
```
본컴포넌트에서 연산하는 복잡한 로직을 별도의 custom 훅을 작성해야 관리하면 좋을 것 같습니다.

hyokualexkwon commented on packages/client/views/SearchedResultsTags/index.tsx:
```diff
@@ -0,0 +1,106 @@
+  optionMap.set('all', '모두');
+  optionMap.set('videos', '영상');
+  optionMap.set('users', '사용자');
+  optionMap.set('tags', '태그');
```
상수의 값을 사용해야 할 것 같습니다.

hyokualexkwon commented on packages/client/views/SearchedResultsUsers/index.tsx:
```diff
@@ -0,0 +1,113 @@
+import { endpoint, searchOptions } from '../../constants';
+import { useSearchUsers } from '../SearchedResults/hooks';
+
+const SearchedResultsUsers: React.FunctionComponent = () => {
```
본 컴포넌트의 복잡한 로직을 별도의 hooks로 관리하면 좋을것 같습니다.

KimJye committed:
[client] | refactor(constants): 새로운 상수객체

기존 배열로 정의해둔 객체를 분리했습니다.

KimJye committed:
[client] | refactor(searched): 상수화로 변경

KimJye committed:
[client] | refactor(search): context api로 변경

전역 상태를 query string으로 관리했었는데 context api로 변경했습니다.

KimJye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/searched_ui_20191203

KimJye committed:
[client] | refactor(search): Context API를 이용하여 상태 관리

KimJye committed:
[client] | refactor(): magic number 상수화

KimJye committed:
[client] | fix(searchedArea): a 태그 관련 수정

빈공간을 클릭해도 a 태그를 누른 효과가 나타났습니다. 그래서
검색 결과의 각 항목의 선택 가능 영역을 컨텐츠 영역으로 제한했습니다.

KimJye committed:
[client] | refactor(viewMore): 더보기 라인 길이 조절

KimJye committed:
[client] | fix(search): 모두 탭을 눌렀을때 404 에러 처리

모두 탭의 url은 search?keyword=java 이런식으로 되기때문에 따로 분기
처리를 해주었습니다.

KimJye committed:
[client] | refactor(searchedResults.hooks): early return 적용

KimJye committed:
[client] | refactor(searchedResultsTitle): 불필요한 Fragment 제거

hyokualexkwon commented on packages/client/components/SearchBar/index.tsx:
```diff
@@ -16,11 +16,20 @@ export const SearchBar: React.FunctionComponent<SearchBarProps> = ({
+    if (e.target.value !== ' ') {
+      setInputValue(e.target.value);
+    }
+    setInputValue(e.target.value.replace(/\s/, ''));
```
이건 어떤 역할을 수행하는 건가요?

KimJye replied:
검색어 입력에 띄어쓰기는 가능해야된다고 생각합니다.
그래서 띄어쓰기의 입력은 수용하되 요청을 보낼때 띄어쓰기를 붙여서 요청합니다.
예를들어, ja va를 입력하게되면 java로 검색하게 됩니다.
이 정책은 유튜브 정책입니다.

Atercatus replied:
유튜브 에서는 ja va 로 검색하면 'ja va le' 'ja va de', 'java; 모두 검색이 됩니다. 
이런식으로 공백을 다 지워버리면 ja va le 와 같은 제목을 가진 비디오는 검색이 누락되지 않나요

KimJye replied:
그렇군요,,,,,,,,

hyokualexkwon commented on packages/client/components/SearchResultsTabProvider/index.tsx:
```diff
@@ -0,0 +1,26 @@
+  }
+};
+
+export const SearchedResultsTabProvider = ({ children }) => {
```
SearchedResults 페이지에서 사용할 컨텍스트 provider니까, 이름에서 Tab을 빼면 좋을것 같아요!

hyokualexkwon commented on packages/client/components/SearchResultsTabProvider/index.tsx:
```diff
@@ -0,0 +1,26 @@
+
+const reducer = (state, action) => {
+  switch (action.type) {
+    case 'tabs':
```
여기서 action은 데이터를 변경하라는 명령이 될것 같습니다. 그러면 action의 type은 동사형이 되어야 할것 같은데요? 

Provider의 tabs의 값을 세팅하는 action이라면, set tabs 라는 식의 이름이 더 적합할 것 같습니다. 한번 고민해 주세요.

+ 상수로 관리하면 좋겠네요!

hyokualexkwon commented on packages/client/components/SearchResultsTabProvider/index.tsx:
```diff
@@ -0,0 +1,26 @@
+};
+
+export const SearchedResultsTabProvider = ({ children }) => {
+  const [state, dispatch] = useReducer(reducer, {});
```
또, 기본 state  값에 빈 객체보다는 
```
{
  tabs: ''
}
```

을 기본값으로 주는건 어떨까요?

hyokualexkwon commented on packages/client/components/SearchedResultsArea/styles.tsx:
```diff
@@ -0,0 +1,65 @@
+  margin-bottom: 2rem;
+  color: white;
+  font-size: 2.2rem;
+  font-weight: 700;
```
コンスタント-!

KimJye replied:
한국어를 사랑하즈악!

hyokualexkwon commented on packages/client/components/SearchedResultsArea/styles.tsx:
```diff
@@ -0,0 +1,65 @@
+  height: 3.4rem;
+  line-height: 3.4rem;
+  font-size: 1.6rem;
+  font-weight: 700;
```
コンスタント-!

hyokualexkwon commented on packages/client/components/SearchedResultsArea/styles.tsx:
```diff
@@ -0,0 +1,65 @@
+  align-items: flex-start;
+  color: white;
+  font-size: 1.6rem;
+  font-weight: 700;
```
コンスタント-!

hyokualexkwon commented on packages/client/components/SearchedResultsTitle/styles.tsx:
```diff
@@ -0,0 +1,23 @@
+    margin-left: 1rem;
+    color: white;
+    font-size: 2rem;
+    font-weight: 700;
```
コンスタント-!

hyokualexkwon commented on packages/client/views/SearchedResults/styles.tsx:
```diff
@@ -39,43 +21,51 @@ export const ContainerGrid = styled(MaterialGrid)`
-      fill: white;
-    }
+    font-size: 2rem;
+    font-weight: 700;
```
コンスタント-!

hyokualexkwon commented on packages/client/components/ViewMore/styles.tsx:
```diff
@@ -0,0 +1,29 @@
+  span {
+    color: white;
+    font-size: 1.6rem;
+    font-weight: 700;
```
コンスタント-!

hyokualexkwon commented on packages/client/components/ViewMore/index.tsx:
```diff
@@ -0,0 +1,29 @@
+  return (
+    <S.ViewMore>
+      <Link
+        prefetch={false}
```
이건 왜 하는거에여?

hyokualexkwon commented on packages/client/views/App/index.tsx:
```diff
@@ -7,18 +7,19 @@ import { ClientContextProvider as FetchingProvider } from 'react-fetching-librar
+    const { pathname, queryString, req } = appContext.ctx;
 
-    const props = initProps(pathname);
+    const props = initProps(pathname, queryString);
```
🎱

hyokualexkwon commented on packages/client/views/App/index.tsx:
```diff
@@ -42,28 +43,38 @@ class MyApp extends App<AppProps> {
+      cacheItems,
+      user,
+      pathname,
+      queryString,
```
🎱

KimJye committed:
[client] | refactor(search): 코드리뷰 반영

쓸데없는 코드를 지웠습니다.
magic number를 상수 처리 했습니다.
action type을 동사로 변경
SearchedResultsTabProvider에서 SearchedResultsProvider로 변경했습니다.
기본 state에 빈 객체 대신에 기본값을 주었습니다.

KimJye committed:
[client] | refactor(searchedResults): magic number 상수화

KimJye committed:
[client] | refactor(): 파일 구조 변경

Atercatus commented on packages/client/components/SearchBar/index.tsx:
```diff
@@ -16,11 +16,20 @@ export const SearchBar: React.FunctionComponent<SearchBarProps> = ({
+  const [inputValue, setInputValue] = useState('');
+
+  const handleChange = e => {
+    if (e.target.value !== ' ') {
```
이런 방식의 예외처리보다 `trim`을 사용해서 좌우 공백을 제거한 이후 validation을 추가로 하는게 더 좋아보입니다.

KimJye replied:
하잇

Atercatus commented on packages/client/components/SearchBar/index.tsx:
```diff
@@ -33,7 +42,13 @@ export const SearchBar: React.FunctionComponent<SearchBarProps> = ({
 
       <S.Input isActive={isActive}>
         <SearchSVG width={20} height={20} viewBox="0 0 24 24" />
```
`viewBox="0 0 24 24"1 이 부분은 SVG 안으로 넣어주세요

Atercatus replied:
width 와 height도 styled component 를 사용해서 가리면 더 좋겠네요

Atercatus commented on packages/client/views/App/index.tsx:
```diff
@@ -42,34 +43,37 @@ class MyApp extends App<AppProps> {
         <ThemeProvider theme={theme}>
           <UserProvider user={user}>
-            {this.isNeedFileProvider(pathname) ? (
+            <SearchedResultsProvider>
```
기존의 코드를 지우진 말고 주석으로 남겨주세요

Atercatus commented on packages/client/views/SearchedResults/hook/use_search.ts:
```diff
@@ -0,0 +1,88 @@
+    }
+    if (payload && !error) {
+      setVideoHasData(true);
+      setVideoHasMore(payload.data.length >= 20);
```
`payload.data.length >= 20)` 와 같은 부분은 함수로 분리해서 의미가 명확하게 드러나면 좋을 거 같습니다.

Atercatus commented on packages/client/views/SearchedResults/hook/use_search.ts:
```diff
@@ -0,0 +1,88 @@
+import { useState, useEffect } from 'react';
```
저희는 파일명에 언더바를 넣지 않습니다.

KimJye replied:
따흐흑

Atercatus commented on packages/client/views/SearchedResults/action/make_query_search_action.ts:
```diff
@@ -0,0 +1,11 @@
+import { Action } from 'react-fetching-library';
```
저희는 파일명에 언더바를 넣지 않습니다.

KimJye replied:
세상에

Atercatus commented on packages/client/views/SearchedResults/index.tsx:
```diff
@@ -0,0 +1,189 @@
+  const router = useRouter();
+  const searchKeyword = router.query.keyword;
+
+  const page = null;
```
왜 사용된 변수 인가요?

KimJye replied:
page가 없는 경우가 있기 떄문입니다~

Atercatus commented on packages/client/views/SearchedResults/index.tsx:
```diff
@@ -0,0 +1,189 @@
+
+  const allCount = hasData ? videoCount + userCount + tagCount : null;
+
+  const optionMap = new Map();
```
다른 곳에서 선언해서 가져오는 것이 좋을 거 같습니다.

Atercatus commented on packages/client/views/SearchedResults/index.tsx:
```diff
@@ -0,0 +1,189 @@
+
+  const customSearchOptions = [];
+
+  if (hasData) {
```
로직은 hook으로 숨기면 좋을 거 같아요

Atercatus commented on packages/client/views/SearchedResults/index.tsx:
```diff
@@ -0,0 +1,189 @@
+          <Grid item xs={12} md={8}>
+            <SearchedResultsTitle searchKeyword={searchKeyword} />
+
+            {allCount === 0 && hasData ? (
```
`hasData` 가 true 인데 검색 결과가 없는 페이지가 노출되는데 계획대로 구현된게 맞나요?

KimJye replied:
네 저희는 여러개의 fetch를 보냅니다. hasData는 다 fetch가 되었을때 true가 됩니다. 그래서 다 fetch가 되었고 fetch한 결과의 개수가 0일때 검색 결과가 없는 페이지가 노출됩니다.
이렇게하므로써 etching이 되고있을때 검색 결과가 없는 페이지가 노출된다는 에러를 잡을 수 있었습니다.

Atercatus commented on packages/client/views/SearchedResults/index.tsx:
```diff
@@ -0,0 +1,189 @@
+                <S.Line />
+
+                {videoCount > 0 && (
+                  <>
```
각각을 컴포넌트로 분리해야 좋을 거 같습니다. 추후에 수정하도록 하죠

Atercatus commented on packages/client/views/SearchedResultsTags/index.tsx:
```diff
@@ -0,0 +1,100 @@
+
+  const { tags, tagHasMore, tagHasData } = useSearchTags(page, searchKeyword);
+
+  const optionMap = new Map();
```
이런 초기화 과정은 외부에서 숨겨주세요

KimJye committed:
[client] | refactor(search): 코드리뷰 반영

검색 공백에대한 피드백을 반영했습니다.
로직을 hook으로 관리하도록 변경했습니다.
svg의 값을 props로 관리하지않고 styled.component로 변경했습니다.
optionMap을 외부에서 선언하고 호출하도록 변경했습니다.

KimJye committed:
[client] | refactor(): 더보기에대한 오류 수정

5개 이하일때도 더보기가 나오는 오류를 수정했습니다.

KimJye committed:
[client] | refactor(search): useRouter 코드 수정

hook이 아닌 helper로 파일을옮겼습니다.

hyokualexkwon commented on packages/client/components/SearchedResultsArea/index.tsx:
```diff
@@ -0,0 +1,69 @@
+            href={`/${SEARCH_OPTION_VALUES.tags}/[tagId]`}
+            as={`/${SEARCH_OPTION_VALUES.tags}/${tag.id}`}
+          >
+            <a onClick={e => e.stopPropagation()}>
```
여기에서 propagation을 막는 이유가 뭐에요?

hyokualexkwon commented on packages/client/views/SearchedResults/helper/get-option-map.ts:
```diff
@@ -0,0 +1,12 @@
+import { SEARCH_OPTION_LABELS, SEARCH_OPTION_VALUES } from '../../../constants';
```
좋네요!

KimJye replied:
ㅎㅎㅎㅎㅎ

hyokualexkwon commented on packages/client/views/SearchedResults/helper/get-option-map.ts:
```diff
@@ -0,0 +1,12 @@
+import { SEARCH_OPTION_LABELS, SEARCH_OPTION_VALUES } from '../../../constants';
+
+export const getOptionMap = () => {
```
makeOptionMap이라는 이름은 어떨까요?

KimJye replied:
반영했습니다아

hyokualexkwon commented on packages/client/views/SearchedResults/hook/use-search.ts:
```diff
@@ -0,0 +1,138 @@
+import { SEARCH_OPTION_LABELS, SEARCH_OPTION_VALUES } from '../../../constants';
+import { makeQuerySearchAction } from '../action/make-query-search-action';
+
+const optionMap = new Map();
```
여기의 optionMap은 별도 helper가 없나봐요!

KimJye replied:
수정했습니당

hyokualexkwon commented on packages/client/views/SearchedResults/hook/use-search.ts:
```diff
@@ -0,0 +1,138 @@
+  return customSearchOptions;
+};
+
+export const makeOptions = optionArray => {
```
helper로 분리해도 좋아보이네요 ㅎㅎ

KimJye replied:
반영했습니당

hyokualexkwon commented on packages/client/views/SearchedResults/hook/use-search.ts:
```diff
@@ -0,0 +1,138 @@
+  return options;
+};
+
+export const useSearchVideos = (page, keyword) => {
```
hook 별로 별도의 파일로 분리하는건 어떠실까요?

KimJye replied:
수정했습니당

KimJye committed:
[client] | refactor(searchedResultsArea): 불필요한 propagation 삭제

KimJye committed:
[client] | refactor(make-option-map): 파일명 변경

get-option-map에서 make-option-map으로 변경했습니다.

KimJye committed:
[client] | refactor(use-search): map을 helper로 분리

KimJye committed:
[client] | refactor(searchedResults): makeOptions를 helper로 분리

KimJye committed:
[client] | refactor(use-search): hook 내용들을 별도의 파일로 분리

hyokualexkwon commented:
오호! 좋네요!

Atercatus commented on packages/client/components/SearchBar/index.tsx:
```diff
@@ -16,11 +16,19 @@ export const SearchBar: React.FunctionComponent<SearchBarProps> = ({
+  const [inputValue, setInputValue] = useState('');
+
+  const handleChange = e => {
+    if (e.target.value.trim()) {
```
`e.target.value`를 따로 변수를 사용해서 옮겨써주세여

Atercatus commented:
빠른 시일 내에 Restructuring 이 필요한거 같습니다..
